### PR TITLE
feat:[CDS-53484]: moving freeze activation handler to redis batch

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
The changes are related to freeze activation handler which sends notification when freeze gets active

Tests Done:-
1. sending notification when freeze scheduled at some time in future
2. sending notification when freeze scheduled recursively

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47114)

### CODE REVIEW

**Review by abhinavmittal-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 6 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47114)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/47114#-NU_IJR72nXzUWaEsRqD:-NU_IJR80wnFZvOGpsDy:bvuooax) ([raw file](https://github.com/harness/harness-core/blob/695a2503f0e2bb13e875ab0379ed4f5a09fb48b4/120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java#L104)):*
> ```Java
>         entity.setShouldSendNotification(false);
>         entity.setNextIteration(null);
>         freezeCRUDService.updateExistingFreezeConfigEntity(entity);
> ```

When we update does this add a value to Audit trails
___
*[`125-freeze/src/main/java/io/harness/freeze/entity/FreezeConfigEntity.java` line 110 at r2](https://reviewable.io/reviews/harness/harness-core/47114#-NU_K97L7r-njuG4K9O0:-NU_K97NAv9H2G8Z9iNh:bltmmh0) ([raw file](https://github.com/harness/harness-core/blob/695a2503f0e2bb13e875ab0379ed4f5a09fb48b4/125-freeze/src/main/java/io/harness/freeze/entity/FreezeConfigEntity.java#L110)):*
> ```Java
>   @SchemaIgnore @LastModifiedBy private EmbeddedUser lastUpdatedBy;
> 
>   Long nextIteration;
> ```

We should not be updating type of entity as it might break older configs

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by abhinavmittal-harness — DISMISSED** [review]
> Reviewed 1 of 6 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47114)* status: 7 file(s) to review. 1 of 2 approvals obtained. 

___
*[`125-freeze/src/main/java/io/harness/freeze/entity/FreezeConfigEntity.java` line 112 at r3](https://reviewable.io/reviews/harness/harness-core/47114#-NUaMiEM-_YHV5pu7iOB:-NUaMiEN6AXX0G61av-_:b-mdeapw) ([raw file](https://github.com/harness/harness-core/blob/aa3bf7c4d15fb28bb513c99a31ea5183ff1c4568/125-freeze/src/main/java/io/harness/freeze/entity/FreezeConfigEntity.java#L112)):*
> ```Java
>   Long nextIteration;
> 
>   List<Long> nextIterations;
> ```

nit: mark this deprecated

**Review by abhinavmittal-harness — DISMISSED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47114)* status: 7 file(s) to review. 1 of 2 approvals obtained.

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java [hunk 1]
```diff
@@ -9,65 +9,78 @@
...
+import java.util.Date;
+import java.util.List;
 import java.util.concurrent.Semaphore;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.data.mongodb.core.MongoTemplate;
 
 @Slf4j
 public class NgDeploymentFreezeActivationHandler implements MongoPersistenceIterator.Handler<FreezeConfigEntity> {
-  private static final int POOL_SIZE = 3;
+  private static final int EPOCH_VALUE_FOR_30_MINS_IN_MILLI = 18000000;
```
Comment (RishabhGupta34) [inline 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java:45]: isn't this 300 mins? 
probably we should use some converter maybe like `TimeUnit.MINUTES.toMillis(30)`?

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java [hunk 2]
```diff
@@ -9,65 +9,78 @@
...
+                           .acceptableNoAlertDelay(ofSeconds(60))
+                           .targetInterval(ofDays(1))
+                           .semaphore(new Semaphore(10))
+                           .handler(this)
+                           .persistenceProvider(new SpringPersistenceRequiredProvider<>(mongoTemplate))
+                           .schedulingType(IRREGULAR_SKIP_MISSED)
+                           .filterExpander(q
+                               -> q.addCriteria(where(FreezeConfigEntityKeys.status).is(FreezeStatus.ENABLED))
+                                      .addCriteria(where(FreezeConfigEntityKeys.nextIteration).ne(null))
+                                      .addCriteria(where(FreezeConfigEntityKeys.shouldSendNotification).is(true))));
```
Comment (RishabhGupta34) [inline 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java:83]: is removing this value `.maximumDelayForCheck(ofMinutes(10))` desired? I see it was present previously but not now

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java (hunk 1 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java:45]: I am not using this variable so removing it

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by RishabhGupta34 — DISMISSED** [review]

**Comment by sourabh-awashti** [comment]
trigger review

**Review by abhinavmittal-harness — DISMISSED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47114)* status: 7 file(s) to review. 2. unresolved discussions.

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java [hunk 3]
```diff
@@ -9,65 +9,77 @@
...
+                       NgDeploymentFreezeActivationHandler.class,
+                       MongoPersistenceIterator.<FreezeConfigEntity, SpringFilterExpander>builder()
+                           .mode(PersistenceIterator.ProcessMode.REDIS_BATCH)
+                           .clazz(FreezeConfigEntity.class)
+                           .fieldName(FreezeConfigEntityKeys.nextIteration)
+                           .acceptableNoAlertDelay(ofSeconds(60))
+                           .targetInterval(ofDays(1))
+                           .semaphore(new Semaphore(10))
+                           .handler(this)
+                           .persistenceProvider(new SpringPersistenceRequiredProvider<>(mongoTemplate))
```
Comment (Raghav-Murali) [inline 120-ng-manager/src/main/java/io/harness/ng/core/handler/freezeHandlers/NgDeploymentFreezeActivationHandler.java:77]: I remember we discussing this on a call sometime back. Redis Batch mode currently supports MorphiaPersistenceProvider only and not SpringPersistenceProvider. 

I had already added and tested the support for SpringPersistenceProvider in this PR which is not merged - https://github.com/harness/harness-core/pull/45505/files#diff-15de57e84344b08335ecc8790cde1947521f9d90e218eeee7f55001a7aea861f. Can you being in the changes of the 2 SpringPersistenceProvider files into this PR ?

**Review by Raghav-Murali — CHANGES_REQUESTED** [review]
Have reviewed only the iterator parts of the code.

**Review by Raghav-Murali — APPROVED** [review]

**Review by abhinavmittal-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47114)* status: Some of the checks are too old. 7 file(s) to review. 2. unresolved discussions.

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**